File: ingestion/json/cxpcFolder.ts
import { readFile, writeFile, mkdir, rename, readdir } from 'node:fs/promises';
import { join, basename } from 'node:path';

async function readJsonArray(filePath: any) {
    try {
        const data = await readFile(filePath, 'utf-8');
        const jsonArray = JSON.parse(data);

        if (!Array.isArray(jsonArray)) {
            throw new Error('The JSON file does not contain a top-level array.');
        }

        console.log(`Loaded ${jsonArray.length} items successfully from ${basename(filePath)}!`);
        return jsonArray;
    } catch (error: any) {
        console.error(`Failed to read or parse JSON file ${basename(filePath)}:`, error.message);
    }
}

function getTodayDateFileName(filename: string, extension: string) {
    const today = new Date();

    const day = String(today.getDate()).padStart(2, '0');
    const month = String(today.getMonth() + 1).padStart(2, '0'); // Months are 0-indexed
    const year = today.getFullYear();

    const ddMMyyyy = `${day}${month}${year}`;
    return filename + "_" + ddMMyyyy + extension;
}

function LambdaCreateTable(data: any, tablename: string) {
    const columns: string[] = Object.keys(data[0])
    const rowValues = Object.values(data[0]);
    let tuple: [string, string, any];
    let arrayOfTuple: [string, string, any][] = [];

    for (let i = 0; i < columns.length; i++) {
        if (typeof rowValues[i] == 'string') {
            tuple = [columns[i], "varchar(20)", rowValues[i]]
        } else if (typeof rowValues[i] == 'number') {
            tuple = [columns[i], "numeric(15, 2)", rowValues[i]]
        } else {
            tuple = [columns[i], "varchar(20)", rowValues[i]]
        }
        arrayOfTuple.push(tuple);
    }

    let createTable = `CREATE TABLE IF NOT EXISTS raw.` + tablename + `\n(\n`
    for (let i = 0; i < arrayOfTuple.length; i++) {
        createTable += '"' + arrayOfTuple[i][0] + '"      ' + arrayOfTuple[i][1] + ',\n'
    }
    createTable = createTable.slice(0, -2) + '\n);';
    return createTable;
}

function LambdaInsertTable(data: any, tablename: string) {
    const lengthOfData = data.length
    let insert = `TRUNCATE TABLE raw.${tablename};\n\n`;

    for (let i = 0; i < lengthOfData; i++) {
        const columns: string[] = Object.keys(data[i])
        const rowValues = Object.values(data[i]);
        let tuple: [string, string, any];
        let arrayOfTuple: [string, string, any][] = [];

        for (let j = 0; j < columns.length; j++) {
            if (typeof rowValues[j] == 'string') {
                tuple = [columns[j], "varchar(20)", rowValues[j]]
            } else if (typeof rowValues[j] == 'number') {
                tuple = [columns[j], "numeric(15, 2)", rowValues[j]]
            } else {
                tuple = [columns[j], "varchar(20)", rowValues[j]]
            }
            arrayOfTuple.push(tuple);
        }

        if (i == 0) {
            let insertParenthesis = "("
            for (let k = 0; k < arrayOfTuple.length; k++) {
                insertParenthesis += '"' + arrayOfTuple[k][0] + '"' + ","
            }
            insertParenthesis = insertParenthesis.slice(0, -1) + ")\nVALUES "

            insert += `INSERT INTO raw.${tablename} ` + insertParenthesis
        }

        insert += "("
        for (let k = 0; k < arrayOfTuple.length; k++) {
            if (typeof arrayOfTuple[k][2] == 'string') {
                insert += `'${arrayOfTuple[k][2]}'`;
            } else if (typeof arrayOfTuple[k][2] == 'number') {
                insert += `${arrayOfTuple[k][2]}`;
            }
            insert += ","
        }
        insert = insert.slice(0, -1) + "),\n"
    }
    insert = insert.slice(0, -2) + ';';
    return insert;
}

async function main() {
    const inputFolder = process.argv[2] || '/Users/jonathankee/Data-Science-Projects/ingestion/sources_unprocessed';
    const outputFolder = '/Users/jonathankee/Data-Science-Projects/ingestion/sql';
    const processedFolder = '/Users/jonathankee/Data-Science-Projects/ingestion/sources_processed';

    try {
        const files = await readdir(inputFolder);
        // Filter files matching the date-suffixed naming convention (e.g., AFR.AI1_16082026.json)
        const targetFiles = files.filter(file => /^.*AI1_\d{8}\.json$/.test(file));

        if (targetFiles.length === 0) {
            console.log('No files matching the pattern *AI1_DDMMYYYY.json found in the directory.');
            return;
        }

        for (const file of targetFiles) {
            const jsonFilePath = join(inputFolder, file);
            
            // Derive generic base name (e.g., AFR.AI1_16082026)
            const fileNameGeneric: string = basename(file, '.json');

            // 1. Transform table name: cxpc_afr_ai1_raw, cxpc_al_ai1_raw, etc.
            const baseWithoutDate = fileNameGeneric.split('_')[0]; // e.g. AFR.AI1
            const tableName = `cxpc_${baseWithoutDate.replace('.', '_')}_raw`.toLowerCase();

            // 2. Transform output SQL filename format: cxpc_BFR_AI1_${TODAY}, cxpc_RGO_AI1_${TODAY}, etc.
            // Extracts prefix (e.g., AFR) and date portion (e.g., 16082026) from the input file
            const [filePrefix, fileDate] = fileNameGeneric.split('_');
            const filePrefixClean = filePrefix.split('.')[0]; // Gets 'AFR' from 'AFR.AI1'
            const customOutputName = `cxpc_${filePrefixClean}_AI1_${fileDate}`;

            // Read file on disk
            let data = await readJsonArray(jsonFilePath);
            if (!data) continue;

            let createTable: string = LambdaCreateTable(data, tableName);
            let insertTable: string = LambdaInsertTable(data, tableName);

            let fullsql = createTable + "\n\n" + insertTable;
            console.log(fullsql);

            // --- Output section ---
            const outputPath = join(outputFolder, `${customOutputName}.sql`);
            const processedFilePath = join(processedFolder, file);

            try {
                // 1. Ensure the SQL output folder exists and write the SQL file
                await mkdir(outputFolder, { recursive: true });
                await writeFile(outputPath, fullsql, 'utf-8');
                console.log(`Successfully saved SQL file to: ${outputPath}`);

                // 2. Ensure the processed folder exists and move the JSON file
                await mkdir(processedFolder, { recursive: true });
                await rename(jsonFilePath, processedFilePath);
                console.log(`Successfully moved JSON file to: ${processedFilePath}`);
            } catch (error: any) {
                console.error(`Error during output or file moving step for ${file}:`, error.message);
            }
        }
    } catch (error: any) {
        console.error('Error reading input directory:', error.message);
    }
}

main();

File: ingestion/json/cxpcFolder.py
import json
import os
import re
from pathlib import Path
import shutil

import pandas as pd
from sqlalchemy import create_engine, text, inspect

# This code implements a latest state pattern.
def process_files():
    # Define directories
    input_folder = Path(
        os.getenv(
            "INPUT_FOLDER",
            "/Users/jonathankee/Data-Science-Projects/ingestion/sources_unprocessed",
        )
    )
    processed_folder = Path(
        "/Users/jonathankee/Data-Science-Projects/ingestion/sources_processed"
    )

    if not input_folder.exists():
        print(f"Input directory does not exist: {input_folder}")
        return

    # Regex pattern matching *AI1_DDMMYYYY.json
    pattern = re.compile(r"^.*AI1_\d{8}\.json$")
    
    # Find target files (No list comprehension)
    target_files = []
    for f in input_folder.iterdir():
        if pattern.match(f.name):
            target_files.append(f)

    if len(target_files) == 0:
        print("No files matching the pattern *AI1_DDMMYYYY.json found in the directory.")
        return

    # Initialize SQLAlchemy Engine
    engine = create_engine("postgresql://postgres:abc123@localhost:5432/prosperous_universe")

    # Ensure the target schema exists
    with engine.begin() as conn:
        conn.execute(text("CREATE SCHEMA IF NOT EXISTS raw;"))

    for file_path in target_files:
        file_name_generic = file_path.stem  # Filename without extension
        parts = file_name_generic.split("_")

        if len(parts) < 2:
            print(f"Skipping malformed filename: {file_path.name}")
            continue

        file_prefix = parts[0]
        file_date_str = parts[1]

        # 1. Transform table name: cxpc_afr_ai1_raw
        table_name = f"cxpc_{file_prefix.replace('.', '_')}_raw".lower()
        staging_table = f"{table_name}_stg"
        primary_keys = ["Interval", "DateEpochMs"]

        # 2. Convert DDMMYYYY (e.g. 16082026) to SQL Date format YYYY-MM-DD (2026-08-16)
        try:
            day = file_date_str[0:2]
            month = file_date_str[2:4]
            year = file_date_str[4:8]
            formatted_date = f"{year}-{month}-{day}"
        except IndexError:
            print(f"Invalid date format in filename: {file_path.name}")
            continue

        # Read JSON file on disk
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data = json.load(f)

            if not isinstance(data, list):
                print(f"The JSON file {file_path.name} does not contain a top-level array.")
                continue

            print(f"Loaded {len(data)} items successfully from {file_path.name}!")
        except Exception as e:
            print(f"Failed to read or parse JSON file {file_path.name}: {e}")
            continue

        # Load into Pandas DataFrame
        df = pd.DataFrame(data)

        # Push metadata and deduplicate (prevents within-batch upsert errors)
        df["file_date"] = formatted_date
        
        # Drop exact duplicate primary keys within the same file to prevent Postgres errors
        df = df.drop_duplicates(subset=primary_keys, keep="last")

        try:
            # Check if dynamic table already exists
            inspector = inspect(engine)
            table_exists = inspector.has_table(table_name, schema="raw")
            
            # Step 1: If target table doesn't exist, create an empty one and set the primary keys
            if not table_exists:
                df.head(0).to_sql(table_name, engine, schema="raw", if_exists="fail", index=False)
                with engine.begin() as conn:
                    # Note: We quote the column names because PostgreSQL forces lowercase unquoted identifiers
                    conn.execute(text(
                        f'ALTER TABLE raw.{table_name} ADD PRIMARY KEY ("Interval", "DateEpochMs");'
                    ))

            # Step 2: Push data to a staging table
            # We use a single connection for the staging table write and merge execution
            with engine.begin() as conn:
                df.to_sql(
                    name=staging_table,
                    con=conn,
                    schema="raw",
                    if_exists="replace", # Replaces the staging table for every batch
                    index=False
                )

                # Step 3: Construct the Merge/Upsert SQL Query
                
                # Gather column names with quotes (No list comprehension)
                columns_quoted = []
                for col in df.columns:
                    columns_quoted.append(f'"{col}"')
                col_str = ", ".join(columns_quoted)
                
                # Gather update conditions (No list comprehension)
                update_clauses = []
                for col in df.columns:
                    if col not in primary_keys:
                        update_clauses.append(f'"{col}" = EXCLUDED."{col}"')
                
                # Gather primary keys with quotes
                pk_quoted = []
                for pk in primary_keys:
                    pk_quoted.append(f'"{pk}"')
                pk_str = ", ".join(pk_quoted)

                if len(update_clauses) > 0:
                    set_clause = ", ".join(update_clauses)
                    conflict_action = f"DO UPDATE SET {set_clause}"
                else:
                    conflict_action = "DO NOTHING"

                # Standard SQL Merge Pattern via INSERT ... ON CONFLICT
                merge_sql = f"""
                    INSERT INTO raw.{table_name} ({col_str})
                    SELECT {col_str} FROM raw.{staging_table}
                    ON CONFLICT ({pk_str})
                    {conflict_action};
                """
                
                # Step 4: Execute the merge and drop the staging table
                conn.execute(text(merge_sql))
                conn.execute(text(f"DROP TABLE raw.{staging_table};"))
            
            print(f"Successfully merged/upserted data into table raw.{table_name}!")

            # Move processed file to processed folder
            processed_folder.mkdir(parents=True, exist_ok=True)
            processed_file_path = processed_folder / file_path.name
            shutil.move(str(file_path), str(processed_file_path))
            print(f"Successfully moved JSON file to: {processed_file_path}\n")

        except Exception as e:
            print(f"Error during database ingestion or file moving for {file_path.name}: {e}\n")


if __name__ == "__main__":
    process_files()